File: src/lib/api/explorer.ts
import { invoke } from "@tauri-apps/api/core";
import type { ExplorerServerView, RepoCarExport, ResolvedExplorerInput, TempBlobFile } from "./types/explorer";

async function resolveInput(input: string): Promise<ResolvedExplorerInput> {
  return invoke("resolve_input", { input });
}

async function describeServer(pdsUrl: string): Promise<ExplorerServerView> {
  return invoke("describe_server", { pdsUrl });
}

async function describeRepo(did: string): Promise<Record<string, unknown>> {
  return invoke("describe_repo", { did });
}

async function listRecords(did: string, collection: string, cursor?: string): Promise<Record<string, unknown>> {
  return invoke("list_records", { did, collection, cursor });
}

async function getRecord(did: string, collection: string, rkey: string): Promise<Record<string, unknown>> {
  return invoke("get_record", { did, collection, rkey });
}

async function exportRepoCar(did: string): Promise<RepoCarExport> {
  return invoke("export_repo_car", { did });
}

async function fetchBlobToTempFile(did: string, cid: string, extension?: string | null): Promise<TempBlobFile> {
  return invoke("fetch_blob_to_temp_file", { cid, did, extension: extension ?? null });
}

async function deleteBlobTempFile(path: string): Promise<void> {
  return invoke("delete_blob_temp_file", { path });
}

async function queryLabels(uri: string): Promise<Record<string, unknown>> {
  return invoke("query_labels", { uri });
}

async function getLexiconFavicons(collections: string[]): Promise<Record<string, string | null>> {
  return invoke("get_lexicon_favicons", { collections });
}

async function clearLexiconFaviconCache(): Promise<void> {
  return invoke("clear_lexicon_favicon_cache");
}

export const ExplorerController = {
  resolveInput,
  describeServer,
  describeRepo,
  listRecords,
  getRecord,
  exportRepoCar,
  fetchBlobToTempFile,
  deleteBlobTempFile,
  queryLabels,
  getLexiconFavicons,
  clearLexiconFaviconCache,
};

File: src-tauri/src/commands/explorer.rs
use crate::error::AppError;
use crate::explorer;
use std::collections::HashMap;
use tauri::AppHandle;

#[tauri::command]
pub async fn resolve_input(input: String) -> Result<explorer::ResolvedExplorerInput, AppError> {
    explorer::resolve_input(input).await
}

#[tauri::command]
pub async fn describe_server(pds_url: String) -> Result<explorer::ExplorerServerView, AppError> {
    explorer::describe_server(pds_url).await
}

#[tauri::command]
pub async fn describe_repo(did: String) -> Result<serde_json::Value, AppError> {
    explorer::describe_repo(did).await
}

#[tauri::command]
pub async fn list_records(
    did: String, collection: String, cursor: Option<String>,
) -> Result<serde_json::Value, AppError> {
    explorer::list_records(did, collection, cursor).await
}

#[tauri::command]
pub async fn get_record(did: String, collection: String, rkey: String) -> Result<serde_json::Value, AppError> {
    explorer::get_record(did, collection, rkey).await
}

#[tauri::command]
pub async fn export_repo_car(did: String, app: AppHandle) -> Result<explorer::RepoCarExport, AppError> {
    explorer::export_repo_car(did, &app).await
}

#[tauri::command]
pub async fn query_labels(uri: String) -> Result<serde_json::Value, AppError> {
    explorer::query_labels(uri).await
}

#[tauri::command]
pub async fn get_lexicon_favicons(
    collections: Vec<String>, app: AppHandle,
) -> Result<HashMap<String, Option<String>>, AppError> {
    explorer::get_lexicon_favicons(collections, &app).await
}

#[tauri::command]
pub async fn clear_lexicon_favicon_cache(app: AppHandle) -> Result<(), AppError> {
    explorer::clear_lexicon_favicon_cache(&app)
}
